### DISCUSSION TITLE
Export Diagram SVG at Server Side

### DISCUSSION THREAD

**Discussion by giuliano-marinelli** [eclipse-glsp/glsp#1439] [Q&A] [answered]
Hello,

Is there a way to export SVG of the edited diagram at server side?
What I want to do is to obtain the visualized diagram SVG when the **saveSourceModel** action is called, so can save a preview of this model in the database. If it can be done on server side it will be more secure for my application.

I'm currently working with a standalone implementation based on node.js
Maybe it's necessary to run a **glsp-client** on the server for render the diagram and then export svg? Or it's possible from **glsp-server** to do this?

If there's no direct way to do this, can you guide me from where I can start?

Thanks,
Cheers

**Comment by martin-fleck-at** [comment, marked as answer]
Hello @giuliano-marinelli,

Currently, the SVG can only be created properly on the client side as all the view rendering (where GModel elements are translated into SVG) is done there. There are already actions that deal with the SVG export, mainly `RequestExportSvgAction` to trigger the SVG export from the client **or the server** and the `ExportSvgAction` which is then sent out with the generated SVG string.

In our standalone workflow example, we use the [standalone export module](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/export/export-modules.ts#L45) to allow the user to hit `Ctrl+Shift+E` to export the current diagram as SVG to a file called `diagram.svg`. The actual saving to a file happens in the [ExportSvgActionHandler](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/export/export-svg-action-handler.ts#L27) which handles the `ExportSvgAction`. 

I'd assume that you can do something similar:
- Send a `RequestExportSvgAction` from the server to the client when you save the source model.
- Register an action handler on the server for the `ExportSvgAction` using the `configureActionHandlers` method of your diagram module.
- In the action handler, you can save the textual SVG code into the database.

**Comment by giuliano-marinelli** [comment]
Hi @martin-fleck-at
Thanks for your answer!

I was trying to instantiate a GLSP client in server side but can't do it because of webpack css imports (I'm not using webpack in backend).
So finally decide to use some of your answer features. Particularly extend the `GLSPSvgExporter`  for get the SVG the way I want, and use it with a custom `SaveModelAction` that allows me to send and extra param for the SVG.

**Comment by arcanefoam** [comment]
For future reference, we implemented this with a Java server in a Eclispe IDE solution. Our issue was that the default GLSP SVG export relies on the client save the svg worked 'okish' in windows (with the issue that the svg is always saved in the Downloads folder), but failed on MacOS where the browser would show the svg instead of saving it.  Our solution was to ask the client to generate the SVG, but instead of saving it through the browser, the client sends the svg contents back to the server, where we can store them. This allowed us to chose the file format (e.g. we produce PDFs instead of raw SVGs) and let the user select the file location.

### Server dispatch

In Eclipse, toolbar/context menus and keyboard shortcuts execute in the IDE, so we need to create 'eclipse.ui' Commands and Handlers, and bind them through the plugin.xml.

```xml
...
  <extension
      point="org.eclipse.ui.commands">
    <command
        categoryId="ca.mcmaster.workflowplus.glsp"
        defaultHandler="org.eclipse.glsp.ide.editor.handlers.ExportDiagramHandler"
        description="Export this diagram to SVG"
        id="ca.mcmaster.workflowplus.glsp.diagram.export"
        name="Export Diagram">
    </command>      
    ...
```
Note that we are using the ExportDiagramHandler that is provided by GLSP.

### Client Handlers

 In the client, we provide our custom SVG exporter, via custom handler (to override the GLSP export diagram handler) and a custom action (that sends the svg instead of saving it).

Our action handler simply dispatches the ExportPdfAction that goes back to the server with the SVG contents.

```ts title="export-handler.ts"
@injectable()
export class ServerExportSvgActionHandler implements IActionHandler {
	
	@inject(TYPES.IActionDispatcher) protected actionDispatcher: GLSPActionDispatcher;
    
	handle(action: Action): void {
		if (ExportSvgAction.is(action)) {
			this.actionDispatcher.dispatch(ExportPdfAction.create(action.svg));
		}
	}
}
```

```ts tittle="export-actions.ts"
export interface ExportPdfAction extends Action {
    kind: typeof ExportPdfAction.KIND;
    svg: string;
}

export namespace ExportPdfAction {
    export const KIND = 'exportPdf';

    export function is(object: unknown): object is ExportPdfAction {
        return Action.hasKind(object, KIND) && hasStringProp(object, 'svg');
    }

    export function create( svg: string ): ExportPdfAction {
        return {
            kind: KIND,
            svg
        };
    }
}
```

```ts tittle="export-module.ts"
export const serverExportModule = new FeatureModule(
    (bind, _unbind, isBound) => {
        const context = { bind, isBound };
        bindAsService(context, TYPES.KeyListener, ExportSvgKeyListener);
        bind(ServerExportSvgActionHandler).toSelf().inSingletonScope();
        configureActionHandler(context, ExportSvgAction.KIND, ServerExportSvgActionHandler);
    },
    { featureId: Symbol('standaloneExport'), requires: exportModule }	
);
```

The final piece of the puzzle is to unload the standard svg export module:

```ts
export function initializeWorkflowplusDiagramContainer(container: Container, ...containerConfiguration: ContainerConfiguration): Container {
    let result = initializeDiagramContainer(
			container,
  ...
    result.unload(standaloneExportModule);
    return result;
}
```

### Server respond

First, we bind our SavePdfHandler in the diagram module

```java
public class MyDiagramModule extends DiagramModule {

  ...
  @Override
  protected void configureActionHandlers(final MultiBinding<ActionHandler> binding) {
    super.configureActionHandlers(binding);
    binding.add(SavePdfHandler.class);
}
  
'''

As discussed previously, the Handler first ask the user for the file location and then starts a Job to generate the PDF.

```java
public class SavePdfHandler extends AbstractActionHandler<SavePdfAction> {

	@Override
	protected List<Action> executeAction(final SavePdfAction actualAction) {
		final String outputPath = this.selectOutputPath();
		if (outputPath == null || outputPath.isBlank()) {
			return List.of();
		}
		final SvgToPdfJob job = new SvgToPdfJob(actualAction.getSvg(), outputPath);
		job.addJobChangeListener(new SvgToPdfJobListener(
			() -> this.showInfoMessage("PDF generated successfully: " + outputPath),
			status -> this.showErrorMessage(status)));
		job.schedule();
		return List.of();
	}

	private void showInfoMessage(final String message) {
		this.showUiMessage(message, true);
	}

	private void showErrorMessage(final IStatus status) {
		final String message = "PDF generation failed" +
			(status != null && status.getMessage() != null && !status.getMessage().isBlank()
				? ": " + status.getMessage()
				: ".");
		this.showUiMessage(message, false);
	}

	private String selectOutputPath() {
		if (!PlatformUI.isWorkbenchRunning()) {
			return "output.pdf";
		}
		final Display display = PlatformUI.getWorkbench().getDisplay();
		if (display == null || display.isDisposed()) {
			return "output.pdf";
		}
		final String[] selectedPath = new String[1];
		display.syncExec(() -> {
			if (display.isDisposed()) {
				return;
			}
			final Shell shell = display.getActiveShell();
			final FileDialog dialog = new FileDialog(shell, SWT.SAVE);
			dialog.setText("Export PDF");
			dialog.setFileName("output.pdf");
			dialog.setFilterExtensions(new String[] { "*.pdf" });
			dialog.setFilterNames(new String[] { "PDF Files (*.pdf)" });
			selectedPath[0] = dialog.open();
		});
		return selectedPath[0];
	}

	private void showUiMessage(final String message, final boolean success) {
		if (!PlatformUI.isWorkbenchRunning()) {
			System.out.println(message);
			return;
		}
		final Display display = PlatformUI.getWorkbench().getDisplay();
		if (display == null || display.isDisposed()) {
			System.out.println(message);
			return;
		}
		display.asyncExec(() -> {
			if (display.isDisposed()) {
				return;
			}
			final Shell shell = display.getActiveShell();
			if (success) {
				MessageDialog.openInformation(shell, "PDF Export", message);
			} else {
				MessageDialog.openError(shell, "PDF Export", message);
			}
		});
	}

}
```

The generation of the PDF is out of the context of GLSP so I will omit it. Suffice to say that we use SVGUniverse and Apache PdfBox to render it. Note that this approach would allow you to have separate implementations that could render other file formats.
